### PR TITLE
Moving `NposSolution` to frame

### PR DESCRIPTION
**Pull request by georgesdib**
Follow up for #11016

Moving `NposSolution` to `frame-election-provider-support`

Polkadot companion: paritytech/polkadot#5112
skip check-dependent-cumulus

@kianenigma

### CODE REVIEW

File: bin/node/runtime/src/lib.rs [hunk 1]
```diff
@@ -598,7 +598,7 @@ frame_election_provider_support::generate_solution_type!(
 );
 
 parameter_types! {
-	pub MaxNominations: u32 = <NposSolution16 as sp_npos_elections::NposSolution>::LIMIT as u32;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:601]: at this point we can probably remove sp_npos_election from the dependencies.

File: frame/election-provider-support/solution-type/src/single_page.rs [hunk 2]
```diff
@@ -102,7 +102,7 @@ pub(crate) fn generate(def: crate::SolutionDef) -> Result<TokenStream2> {
 		#vis struct #ident { #single #rest }
 
 		use _npos::__OrInvalidIndex;
```
Comment (kianenigma) [inline frame/election-provider-support/solution-type/src/single_page.rs:104]: is this the only pace where this _npos is used?

File: frame/election-provider-support/solution-type/src/single_page.rs [hunk 3]
```diff
@@ -174,9 +174,9 @@ pub(crate) fn generate(def: crate::SolutionDef) -> Result<TokenStream2> {
 		}
 
 		type __IndexAssignment = _npos::IndexAssignment<
-			<#ident as _npos::NposSolution>::VoterIndex,
-			<#ident as _npos::NposSolution>::TargetIndex,
-			<#ident as _npos::NposSolution>::Accuracy,
+			<#ident as frame_election_provider_support::NposSolution>::VoterIndex,
```
Comment (kianenigma) [inline frame/election-provider-support/solution-type/src/single_page.rs:177]: This is wrong. You now assume the existence of this module in scope. 

Look into how _npos is created, we need a similar approach. And ideally we remove _npos as it is nota used much anymore. 

(sorry for the short feedback, hopefully it is clear, afk now)

File: frame/election-provider-support/solution-type/src/single_page.rs (hunk 2 above)
Comment (georgesdib) [reply frame/election-provider-support/solution-type/src/single_page.rs:104]: I removed _npos from everywhere and replaced it with _fpes (not sure if this name is ok as a shorthand for `frame_election_provider_support`?)

File: frame/election-provider-support/solution-type/src/single_page.rs (hunk 3 above)
Comment (georgesdib) [reply frame/election-provider-support/solution-type/src/single_page.rs]: Implemented it better, I had to do a few more changes to completely remove the dependency on _npos

File: bin/node/runtime/src/lib.rs [hunk 4]
```diff
@@ -621,10 +621,12 @@ pub const MINER_MAX_ITERATIONS: u32 = 10;
 
 /// A source of random balance for NposSolver, which is meant to be run by the OCW election miner.
 pub struct OffchainRandomBalancing;
-impl frame_support::pallet_prelude::Get<Option<(usize, sp_npos_elections::ExtendedBalance)>>
-	for OffchainRandomBalancing
+impl
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:624]: I think you are using too much explicit imports: `Get` is already in scope, and `ExtendedBalance` is used twice, so it should be imported once.

File: bin/node/runtime/src/lib.rs [hunk 5]
```diff
@@ -621,10 +621,12 @@ pub const MINER_MAX_ITERATIONS: u32 = 10;
...
+	frame_support::pallet_prelude::Get<
+		Option<(usize, frame_election_provider_support::ExtendedBalance)>,
+	> for OffchainRandomBalancing
 {
-	fn get() -> Option<(usize, sp_npos_elections::ExtendedBalance)> {
+	fn get() -> Option<(usize, frame_election_provider_support::ExtendedBalance)> {
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:629]:
```suggestion
	fn get() -> Option<(usize, ExtendedBalance)> {
```

File: bin/node/runtime/src/lib.rs [hunk 6]
```diff
@@ -693,7 +695,7 @@ impl pallet_bags_list::Config for Runtime {
 	type ScoreProvider = Staking;
 	type WeightInfo = pallet_bags_list::weights::SubstrateWeight<Runtime>;
 	type BagThresholds = BagThresholds;
-	type Score = sp_npos_elections::VoteWeight;
+	type Score = frame_election_provider_support::VoteWeight;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:698]:
```suggestion
	type Score = VoteWeight;
```

File: frame/election-provider-multi-phase/src/unsigned.rs [hunk 7]
```diff
@@ -53,8 +52,8 @@ pub type Assignment<T> =
 	sp_npos_elections::Assignment<<T as frame_system::Config>::AccountId, SolutionAccuracyOf<T>>;
 
 /// The [`IndexAssignment`][sp_npos_elections::IndexAssignment] type specialized for a particular
-/// runtime `T`.
-pub type IndexAssignmentOf<T> = sp_npos_elections::IndexAssignmentOf<SolutionOf<T>>;
+/// runtime `T`.
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/unsigned.rs:55]:
```suggestion
/// runtime `T`.
```

**Review by kianenigma — APPROVED** [review]
Needs to revise some verbose imports, otherwise looks good to me.

**Review by emostov — APPROVED** [review]

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**